[CS2113T-W13-2] Fitr

jerichochua opened the pull request:
Fitr will help you to keep fit and stay healthy by tracking your calories, and recommending exercises based on your schedule, all of which is done through a command-line interface.

esskayzz commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+Figure 2 below shows how the components work with one another, when a user enters the following command `food chicken rice /600 1`:
+
+![Add food sequence diagram](images/AddFoodSequenceDiagram.png)
+
```
Should there be an activation bar for :Fitr as well?

esskayzz commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+![Main class diagram](images/MainClassDiagram.png)
+
+_Figure 1: Architecture diagram of Fitr_
```
I'm not sure if the abstract class command should have an arrow to Fitr

Aliciaho commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+Figure 1 above explains the high-level design of Fitr. Given below is a quick overview of each component.
+
+The main driver of the application is `Fitr` is responsible for mainly two phases:
+* At app launch: Initialises or loads the respective lists and user profile in the correct sequence and is in charge of connecting the various components with each other.
```
Maybe you could elaborate what is the correct sequence is.

elizabethcwt commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+Figure 2 below shows how the components work with one another, when a user enters the following command `food chicken rice /600 1`:
+
+![Add food sequence diagram](images/AddFoodSequenceDiagram.png)
```
Hello! The sequence diagram looks good! One thing I'd like to point out would be that the Fitr class does not display the relevant activation bar(s). Activation bars are optional if self-explanatory, but for the other classes, activation bars have been shown. Hence, I think including the activation bar(s) for the Fitr class would improve consistency! :)

Aliciaho replied:
Maybe you can add a brief explanation for the food sequence diagram to make it clearer?

esskayzz commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+Classes used by multiple components are in the `fitr.common` package.
+
+## 4. Implementation
+
+This section describes how some of the features in Fitr are implemented.
+
```
Implementation part is clearly written, understandable and has good use of diagrams. However, you might want to elaborate more about the functions. For example, you show viewGoal() for 4.3 view, but what are the other commands or fields that I might want view. You might want to add a little more detail on that.

Some paragraphs can be wordy, you might want to break them up into smaller paragraphs.

Aliciaho commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+This component also listens to other components and outputs the desired messages in specified formats.
+
+#### 3.2.2 Parser component
```
Maybe you can try to add sequence diagram here?

Aliciaho commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+The Parser component takes in the user input from the `Ui` class and handles the various commands. If the command is invalid, it calls the `Ui` class to prompt the user until valid commands are entered.
+When a valid command is keyed in, the `Parser` class returns a `Command` object to execute the command.
+
+#### 3.2.3 Command component
```
Maybe you can try adding a class diagram here?

Aliciaho commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+The Command component consists of an abstract `Command` class and the various different commands that inherit the `Command` class.
+Each type of command class (e.g. `addFoodCommand`, `addExerciseCommand` etc) implements an abstract `execute()` method that carries out the command. 
+
+#### 3.2.4 User component
```
Maybe you can try adding a class diagram here?

esskayzz commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+_Figure 2: Sequence diagram when adding a food into Fitr_
+ 
+### 3.2 Components
```
This part is written clearly and succinctly. Especially on the different behaviour when different inputs are given.

elizabethcwt commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+When the user enters an edit command, it first passes through `Parser`. Once the input is parsed as an edit command, it is then passed to `EditCommandParser`, where it further parses the user's input. As the user is able to edit either the individual profile characteristics, food or exercise entries, the `EditCommandParser` is able parse what the user intends to edit. For example, if the user int ...
+Figure 4 below shows the sequence diagram when the user enters `edit exercise 25/10/2020 1 push ups /100`.
+
+![Sequence diagram for editing exercise](images/EditExerciseSequenceDiagram.png)
```
Hi, I've noticed that the activation bar in the Fitr class was not properly closed, despite the edit command being successfully executed. Other than that, the sequence diagram looks good. Keep up the good work! 👍

Aliciaho commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+When the user enters 'help', the `Ui` class reads it and passes it through `Parser`.  Once the input is parsed as a help command, the user input is passed to `HelpCommand`, which calls `Ui` to print the help message.
+Figure 5 below shows the sequence diagram when the user enters `help`.
+
+![Sequence diagram for help command](images/HelpCommandSequenceDiagram.png)
```
The activation bar for the sequence diagram was not properly closed, maybe you would like to edit it?

Aliciaho commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+When the user enters a view command, the `Ui` class reads it and passes it through `Parser`. Once the input is parsed as a view command, it is then passed to `ViewCommand`, where it is handled based on the type of view command. Then the respective view method is called to output the messages via the `Ui` class. For example, if you intend to view your goal entry, the input is passed to `ViewComman ...
+
+![Sequence diagram for view goal command](images/ViewGoalSequenceDiagram.png)
```
I have realised that the sequence diagram here was not properly closed, maybe you would like to edit it?

elizabethcwt commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,243 @@
+
+### 3.1 Architecture
+
+This section illustrates the high-level architecture of Fitr, as well as an illustration of the components used in Fitr.
```
Hello, the architecture diagram looks neat. Perhaps a succinct overview of the diagram would make the entire architecture of Fitr more clear to the reader before they delve into the rest of the Developer Guide! :)

e0426051 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,251 @@
+
+### 3.1 Architecture
+
+This section illustrates the high-level architecture of Fitr, as well as an illustration of the components used in Fitr.
```
Perhaps consider using alternatives instead of Aggregations for Goal, Food and Exercise to improve clarity. Also, without looking at the code, I'm not sure if the Goal is supposed to contain GoalList instead of the other way round(Similar issues for Food/FoodList and Exercise/ExerciseList)

e0426051 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,251 @@
+Figure 1 above explains the high-level design of Fitr. Given below is a quick overview of each component.
+
+The main driver of the application is `Fitr` is responsible for mainly two phases:
+* At app launch: Initialises or loads the respective lists and user profile in the correct sequence and is in charge of connecting the various components with each other.
```
As mentioned by Alicia, this part may be imprecise, and hence perhaps a adding a relevant diagram or a elaboration might help.

e0426051 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,251 @@
+
+![Add food sequence diagram](images/AddFoodSequenceDiagram.png)
+
+_Figure 2: Sequence diagram when adding a food into Fitr_
```
The sequence diagram is generally well done. The instances are evenly spaced which makes it easy for readers to navigate through the diagram. There are also no 'bugs' in the diagram as far as I can tell.

jzihao97 replied:
![image](https://user-images.githubusercontent.com/57383789/97662059-fd3afd80-1ab0-11eb-944b-9fecf6736db3.png)

Consider delay the deletion to make the sequence diagram clearer.

e0426051 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,251 @@
+
+![StorageManager class diagram](images/StorageClassDiagram.png)
+
+_Figure 3: `StorageManager` class diagram_
```
Again, the diagram seems to insinuate that Goal contains GoalList. If this is by design, perhaps consider renaming Goal to GoalContainer or equivalent to avoid potential confusion. (Same for Food and Exercise).

jzihao97 commented on docs/DeveloperGuide.md:
```diff
@@ -1,34 +1,251 @@
+
+Figure 7 below shows the sequence diagram when the user inputs the `clear` command.
+
+![Sequence diagram for clear command](images/ClearCommandSequenceDiagram.png)
```
![image](https://user-images.githubusercontent.com/57383789/97661912-a0d7de00-1ab0-11eb-9756-6e1f6fac8289.png)

Does view command access the variables of other objects, maybe it is possible to add them in.

jzihao97 reviewed (commented):
Overall very nicely organized DG.

hui444 committed:
Merge pull request #182 from jerichochua/branch-BugFixes

Fix bug when invalid data is in user profile text file

jerichochua committed:
Merge pull request #185 from hui444/UpdateGoals-branch

Update goals to use LocalDate

jerichochua committed:
Merge pull request #184 from dmbclub/Mingbang

update viewcommand

jerichochua committed:
Refactor formatter as a static member in DateManager

jerichochua committed:
Add input indicator

dmbclub committed:
Merge pull request #186 from jerichochua/branch-Cleanup

Refactor formatter as a static member in DateManager

HarrisMaung committed:
commits

HarrisMaung committed:
Merge branch 'master' into Updatesreal

HarrisMaung committed:
Maincheck updates

jerichochua committed:
Merge pull request #187 from sixletters/Updatesreal

Updatesreal

gohsonghan98 committed:
Merge branch 'master' into branch-FixEditProfile

gohsonghan98 committed:
Fix name input format

gohsonghan98 committed:
Fix checkstyle error

jerichochua committed:
Add JUnit tests for EditCommandParser class

jerichochua committed:
Add logging under EditEntryCommand class

jerichochua committed:
Merge pull request #179 from gohsonghan98/branch-FixEditProfile

Fix edit profile commands and features

jerichochua committed:
Merge branch 'master' into JUnitTests

jerichochua committed:
Remove duplicated code

hui444 committed:
Merge pull request #189 from jerichochua/branch-Cleanup

Remove duplicated code

Commit:
Change symbols, add JUnit tests for Goal

Commit:
Fix code style

jerichochua committed:
Merge pull request #190 from hui444/changeSymbols-branch

Change symbols, add JUnit tests for Goal

jerichochua committed:
Fix bug when text files contain invalid data

jerichochua committed:
Fix failing checks due to imports

hui444 committed:
Merge pull request #188 from jerichochua/JUnitTests

Make code more defensive

hui444 committed:
Merge pull request #191 from jerichochua/branch-BugFixes

Fix bug when text files contain invalid data

Commit:
Update Ui and DG diagrams

Commit:
Fix code style

jerichochua committed:
Merge pull request #192 from hui444/ErrorMessage-branch

Update Ui and DG diagrams

jerichochua committed:
Rename PPP

jerichochua committed:
Update AboutUs.md

jerichochua committed:
Update images in DG

jerichochua committed:
Merge pull request #193 from jerichochua/UGDG

Update images in DG

HarrisMaung committed:
updates

HarrisMaung committed:
Updates on colour

jerichochua committed:
Merge pull request #194 from sixletters/UGupdates

updates

Commit:
Specify errors, validate help command

Commit:
Fix code style

jerichochua committed:
Add separators between inputs and outputs

jerichochua committed:
Merge branch 'master' into branch-BugFixes

jerichochua committed:
Merge pull request #195 from hui444/updates-branch

Specify errors, validate help command

hui444 committed:
Merge pull request #196 from jerichochua/branch-BugFixes

Add separators between input and output

jerichochua committed:
Change successful edit entry string

jerichochua committed:
Update PPP

hui444 committed:
Merge pull request #197 from jerichochua/branch-Cleanup

Change successful edit entry string

Commit:
Changed separator length, Update UG

Commit:
Add PPP section

Commit:
Add extra parameter validation for view and help command, Add specific error in add food command

hui444 committed:
Merge pull request #198 from jerichochua/jericho-AboutUs

Update PPP

HarrisMaung committed:
updates on recommender

jerichochua committed:
Fix date bug in edit command

HarrisMaung committed:
Update Parser.java

HarrisMaung committed:
updates on checkstye

dmbclub committed:
Merge pull request #201 from sixletters/Bugupdates

Bug and recommender updates

Commit:
Merge branch 'master' into updates-branch

Commit:
Updated help command and Remove extra parameters for mark as complete

Commit:
Fix code style

Commit:
Minor changes

jerichochua committed:
Fix bug when empty name is given as input for edit command

jerichochua committed:
Catch DateTimeParseException in edit command

Commit:
Change view command if statements

gohsonghan98 committed:
Merge pull request #199 from hui444/updates-branch

Updated separator, UG and Add PPP section

jerichochua committed:
Update DG with ListManager component

hui444 committed:
Merge pull request #200 from jerichochua/branch-BugFixes

Fix date bug in edit command

hui444 committed:
Merge pull request #202 from jerichochua/UGDG

Update DG with ListManager component

gohsonghan98 committed:
Improve view command

dmbclub committed:
Merge pull request #20 from AY2021S1-CS2113T-W13-2/master

update for veresion 2.1

hui444 committed:
Merge pull request #203 from gohsonghan98/branch_fix_viewcommand

Improve view command

Commit:
Print create data for goals, Rearrange goal list format, Specify error for view command

Commit:
Change view tips to blue

Commit:
Update UG

Commit:
Remove extra line

dmbclub committed:
Add Junit Tests

dmbclub committed:
PPP

HarrisMaung committed:
Bugupdates and UG updates

sixletters committed:
Merge branch 'master' into Bugupdates

dmbclub committed:
update tips

dmbclub committed:
update PPP

dmbclub committed:
checkstyle

dmbclub committed:
Merge pull request #205 from sixletters/Bugupdates

Bugupdates and UG updates

dmbclub committed:
checkstyle

dmbclub committed:
checkstyle

dmbclub committed:
checkstyle

dmbclub committed:
checkstyle

Commit:
Update comment to blue and PPP template

hui444 committed:
Merge branch 'master' into Update-branch

dmbclub committed:
remove commandstest

jerichochua committed:
Update UG with missing parts

jerichochua committed:
Merge pull request #206 from dmbclub/Mingbang

Junit tests , PPP

Commit:
Merge remote-tracking branch 'upstream/master' into Update-branch

Commit:
Update help message in UG

Commit:
Add tip for adding smart goals and Add specified format error in edit command

Commit:
Add separator in recommend command and printing in goal and view command

dmbclub committed:
Merge pull request #21 from AY2021S1-CS2113T-W13-2/master

update for version 2.1

gohsonghan98 committed:
Update UG for edit profile and view command

gohsonghan98 committed:
Fix weight and height inputs to round up to 2dp

gohsonghan98 committed:
Update height and weight to 2dp

jerichochua committed:
Merge pull request #204 from hui444/Update-branch

Update branch

jerichochua committed:
Update PPP

hui444 committed:
Merge pull request #207 from jerichochua/jericho-AboutUs

Update PPP

gohsonghan98 committed:
Merge branch 'master' into branch-FixInputDecimal

gohsonghan98 committed:
Merge branch 'master' into branch-UserGuideUpdate2

jerichochua committed:
Merge pull request #208 from gohsonghan98/branch-FixInputDecimal

Fix input decimal places

gohsonghan98 committed:
Change notes format for edit commands

gohsonghan98 committed:
Merge branch 'master' into branch-UserGuideUpdate2

jerichochua committed:
Merge pull request #209 from gohsonghan98/branch-UserGuideUpdate

User guide update

HarrisMaung committed:
DGupdates

HarrisMaung committed:
Updates

HarrisMaung committed:
Checkstyle updates

jerichochua committed:
Move exit message to ExitCommand

jerichochua committed:
Fix symbols in tips.txt

jerichochua committed:
Update some messages

jerichochua committed:
Handle case when no entries found on a date while editing

A message is shown indicating that no entries are found on the specified
date.

jerichochua committed:
Edit more messages

jerichochua committed:
Update UG

gohsonghan98 committed:
Add sequence diagram for edit and view commad

gohsonghan98 committed:
Add PPP

gohsonghan98 committed:
Update PPP

gohsonghan98 committed:
Update DG

Commit:
Limit target goal calories, Validate text file goals, More complex validation for smart goals, Remove additional separator after bye command

Commit:
Write to goal list when loaded, Fix code style

Commit:
Update comment

hui444 committed:
Merge pull request #211 from jerichochua/branch-BugFixes

Quick changes

Commit:
Remove additional line

hui444 committed:
Merge pull request #212 from jerichochua/UGDG

Update UG

hui444 committed:
Merge pull request #213 from gohsonghan98/branch-PPP

Update PPP

hui444 committed:
Merge pull request #214 from gohsonghan98/branch-UpdateDG

Update DG

Commit:
Merge remote-tracking branch 'upstream/master' into goals-branch

Commit:
Add to PPP

Commit:
Add to UG

Commit:
Add to .gitignore

jerichochua committed:
Merge pull request #215 from hui444/goals-branch

Limit target goal calorie values, validate txt file goals,..

dmbclub committed:
Merge pull request #22 from AY2021S1-CS2113T-W13-2/master

version 2.1

dmbclub committed:
add link to code

jerichochua committed:
Further updates to UG

dmbclub committed:
Set theme jekyll-theme-cayman

dmbclub committed:
Set theme jekyll-theme-cayman

jerichochua committed:
Merge pull request #216 from dmbclub/Mingbang

Add links

jerichochua committed:
Update UG expected output with separators

jerichochua committed:
Update DG images

jerichochua committed:
Update DG images

HarrisMaung committed:
Update on bugs

sixletters committed:
Merge branch 'master' into DGupdates

HarrisMaung committed:
Update AddFoodCommand.java

HarrisMaung committed:
Merge branch 'DGupdates' of https://github.com/sixletters/tp into DGupdates

HarrisMaung committed:
Create goals.txt

hui444 committed:
Merge pull request #217 from jerichochua/UGDG

Further updates to UG/DG

dmbclub committed:
Merge pull request #23 from AY2021S1-CS2113T-W13-2/master

version 2.1

HarrisMaung committed:
Updates

HarrisMaung committed:
Updates

jerichochua committed:
Validate and skip invalid entries in food list file

HarrisMaung committed:
updates

HarrisMaung committed:
updates

HarrisMaung committed:
updates

dmbclub committed:
Merge pull request #210 from sixletters/DGupdates

DGupdates

dmbclub committed:
Merge pull request #24 from AY2021S1-CS2113T-W13-2/master

updates

jerichochua committed:
Validate and skip invalid entries in exercise list file

Also updated tests.

dmbclub committed:
number the figures

sixletters committed:
Merge pull request #218 from dmbclub/Mingbang

number the figures

jerichochua committed:
Validate user profile

dmbclub committed:
Merge pull request #219 from jerichochua/branch-Cleanup

Validate content in exercise, food, user files

HarrisMaung committed:
DGupdates

sixletters committed:
Merge branch 'master' into FinalDG

sixletters committed:
Merge pull request #220 from sixletters/FinalDG

DGupdates

HarrisMaung committed:
Updates

gohsonghan98 committed:
Merge pull request #221 from sixletters/FinalFinalDG

Updates

jerichochua committed:
Update add food sequence diagram

jerichochua committed:
Update UG, DG

hui444 committed:
Merge pull request #222 from jerichochua/UGDG

Update add food sequence diagram

Commit:
Add validation to date for goal txt file, Add assertion in goal storage

jerichochua committed:
Merge pull request #223 from hui444/ValidateTxtFiles-branch

Add validation to date for goal txt file, Add assertion in goal storage

dmbclub committed:
fix bug

dmbclub committed:
fix bug

jerichochua committed:
Merge pull request #224 from dmbclub/Mingbang

fix bug

Commit:
Change splitting spaces

jerichochua committed:
Merge pull request #225 from hui444/ValidateTxtFiles-branch

Change splitting spaces

Commit:
Change all splitting spaces

dmbclub committed:
Merge pull request #226 from hui444/ValidateTxtFiles-branch

Change all splitting spaces

Commit:
Catch NumberFormatException in formatGoals

gohsonghan98 committed:
Fix User class bugs and improve messages

gohsonghan98 committed:
Fix checkstyle and improve messages

Commit:
Change error message thrown

jerichochua committed:
Merge pull request #227 from gohsonghan98/branch-fixUser

User bug fix and message improvements

HarrisMaung committed:
DGupdates

asdasd

jerichochua committed:
Merge pull request #228 from hui444/Bugs-branch

Catch NumberFormatException in FormatGoals, Change error message thrown

HarrisMaung committed:
UG updates

HarrisMaung committed:
dg updates

Commit:
Add more checks to goal txt file

Commit:
Add more checks to txt file, Refactor remove leading zeros

Commit:
Fix code style

Commit:
Add catch in goalstorage

Commit:
Update JUnit test for goalstorage

HarrisMaung committed:
Logger updates

Commit:
Add to PPP

hui444 committed:
Update hui444.md

hui444 committed:
Update hui444.md

HarrisMaung committed:
Updates on PPP

jerichochua committed:
Merge pull request #229 from sixletters/REALFINALDG

DGupdates

jerichochua committed:
Fix file name

gohsonghan98 committed:
Add logging and assertions

gohsonghan98 committed:
Create User class diagram

jerichochua committed:
Merge pull request #230 from hui444/Bugs-branch

Add more checks to txt file and Refactor removing leading zeros method

gohsonghan98 committed:
Merge branch 'master' into branch-UserClassDiagram

gohsonghan98 committed:
Add User class diagram into DG and relabel diagrams

gohsonghan98 committed:
Merge branch 'master' into branch-LoggingAndAssertions

gohsonghan98 committed:
Update PPP

jerichochua committed:
Merge pull request #231 from gohsonghan98/branch-UserClassDiagram

Add User class diagram

hui444 committed:
Merge pull request #232 from gohsonghan98/branch-LoggingAndAssertions

Logging and assertions

gohsonghan98 committed:
Merge pull request #233 from sixletters/REALFINALDG

Updates on PPP

Commit:
Change error thrown in completegoal command

jerichochua committed:
Fix no validation when editing entries, and more

- Caught exception when deleting with invalid date
- Update limits
- Update messages
- Update UG

Commit:
Fix calories = 0 for goal

jerichochua committed:
Update PPP

jerichochua committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-W13-2/tp into branch-Cleanup

Commit:
Remove additional line

Commit:
Change calculations for smart goal, Add additional information to UG

Commit:
Add special case where targetcalorie = 0

Commit:
Specify errors in delete command

hui444 committed:
Merge pull request #235 from jerichochua/branch-Cleanup

More last minute updates

jerichochua committed:
Merge pull request #234 from hui444/GoalError-branch

Change error thrown in completegoal command

Commit:
Revert delete commands

gohsonghan98 committed:
Merge pull request #236 from hui444/RevertDeleteCommand-branch

Revert delete commands

Commit:
remove lines

jerichochua committed:
Update PPP

gohsonghan98 committed:
Merge pull request #237 from hui444/revert

Remove lines in delete command

Commit:
Catch ArrayIndexOutOfBoundsException in edit command

gohsonghan98 committed:
Fix User setup bug

gohsonghan98 committed:
Merge pull request #238 from hui444/lastbugs

Catch ArrayIndexOutOfBoundsException in edit command

jerichochua committed:
Fix UG mistakes

hui444 committed:
Merge pull request #239 from gohsonghan98/branch-FixUserBug

Fix User setup bug

jerichochua committed:
Update UserGuide.md

Commit:
Fix case in complete goal command

jerichochua committed:
Update UserGuide.md

Commit:
Fix case in delete command

jerichochua committed:
Merge pull request #240 from hui444/lastbugs

case sensitive bugs in complete goal and delete commands

gohsonghan98 committed:
Fix sequence diagram

gohsonghan98 committed:
Fix sequence diagram

jerichochua committed:
Merge pull request #241 from gohsonghan98/branch-FixDG

Fix sequence diagram

Commit:
Update UG

Commit:
Update DG

jerichochua committed:
Update images

jerichochua committed:
Merge pull request #242 from hui444/finalchanges-branch

Update UG

hui444 committed:
Merge pull request #243 from jerichochua/master

Update images

jerichochua committed:
Update add exercise diagram

hui444 committed:
Merge pull request #244 from jerichochua/master

Update add exercise diagram

jerichochua committed:
Update UserGuide.md

jerichochua committed:
Update DeveloperGuide.md

jerichochua committed:
Update architecture diagram

jerichochua committed:
Merge branch 'master' of https://github.com/AY2021S1-CS2113T-W13-2/tp

jerichochua committed:
Merge pull request #245 from jerichochua/master

Update architecture diagram
